### ISSUE TITLE
after migrating to zato 3 from 2.0.7 Plain HTTP outgoing connections do not work

### ISSUE THREAD

**Issue by raybog** [zatosource/zato#770]
Zato 3.0.0rc1+rev.a0d5040d

```
Could not ping the connection, e:[Traceback (most recent call last):
  File "/opt/zato/tmp/zato/code/zato-web-admin/src/zato/admin/web/views/__init__.py", line 519, in id_only_service
    result = req.zato.client.invoke(service, {'id': id})
  File "/opt/zato/tmp/zato/code/zato-web-admin/src/zato/admin/middleware.py", line 78, in invoke
    raise Exception('CID: {}\nDetails: {}'.format(zato_env.get('cid'), zato_env.get('details')))
Exception: CID: f554a92be5802a182654eff2
Details: Traceback (most recent call last):
  File "/opt/zato/tmp/zato/code/zato-server/src/zato/server/connection/http_soap/channel.py", line 212, in dispatch
    payload, worker_store, self.simple_io_config, post_data, path_info, soap_action)
  File "/opt/zato/tmp/zato/code/zato-server/src/zato/server/connection/http_soap/channel.py", line 369, in handle
    params_priority=channel_item.params_pri)
  File "/opt/zato/tmp/zato/code/zato-server/src/zato/server/service/__init__.py", line 457, in update_handle
    self._invoke(service, channel)
  File "/opt/zato/tmp/zato/code/zato-server/src/zato/server/service/__init__.py", line 380, in _invoke
    service.handle()
  File "/opt/zato/tmp/zato/code/zato-server/src/zato/server/service/internal/service.py", line 306, in handle
    response = func(id_, payload, channel, data_format, transport, serialize=True)
  File "/opt/zato/tmp/zato/code/zato-server/src/zato/server/service/__init__.py", line 579, in invoke
    return self.invoke_by_impl_name(self.server.service_store.name_to_impl_name[name], *args, **kwargs)
  File "/opt/zato/tmp/zato/code/zato-server/src/zato/server/service/__init__.py", line 561, in invoke_by_impl_name
    return self.update_handle(*invoke_args, **kwargs)
  File "/opt/zato/tmp/zato/code/zato-server/src/zato/server/service/__init__.py", line 500, in update_handle
    raise Exception(exc_formatted)
Exception: Traceback (most recent call last):
  File "/opt/zato/tmp/zato/code/zato-server/src/zato/server/service/__init__.py", line 457, in update_handle
    self._invoke(service, channel)
  File "/opt/zato/tmp/zato/code/zato-server/src/zato/server/service/__init__.py", line 380, in _invoke
    service.handle()
  File "/opt/zato/tmp/zato/code/zato-server/src/zato/server/service/internal/http_soap.py", line 400, in handle
    self.response.payload.info = config_dict.get(item.name).ping(self.cid)
AttributeError: 'NoneType' object has no attribute 'get'


]
```

**Comment by shizacat** [comment]
A similar problem:

conn = self.outgoing.plain_http.get('cellid.1c_request')

```
2017-05-29 07:14:31,868 - ERROR - 28:DummyThread-4616 - zato.server.connection.http_soap.channel:258 - Caught an exception, cid:`5e5916325bd577febb79b6c3`, status_code:`500`, _format_exc:`Traceback (most recent call last):
  File "/opt/zato/code/zato-server/src/zato/server/connection/http_soap/channel.py", line 212, in dispatch
    payload, worker_store, self.simple_io_config, post_data, path_info, soap_action)
  File "/opt/zato/code/zato-server/src/zato/server/connection/http_soap/channel.py", line 369, in handle
    params_priority=channel_item.params_pri)
  File "/opt/zato/code/zato-server/src/zato/server/service/__init__.py", line 457, in update_handle
    self._invoke(service, channel)
  File "/opt/zato/code/zato-server/src/zato/server/service/__init__.py", line 380, in _invoke
    service.handle()
  File "/opt/zato/code/conf/work/hot-deploy/current/cache_cellid.py", line 208, in handle
    requset_1c = self.outgoing.plain_http.get('cellid.1c_request')
AttributeError: 'NoneType' object has no attribute 'get'
```

**Comment by dsuch** [comment]
Hi @shizacat,

thanks for your pull request over at https://github.com/zatosource/zato/pull/774 but this should look a bit differently. I will explain.

In general, you are correct that this attribute is missing, and I thank you for the PR, but it should be added in a different place, namely here below in ServiceStore.

https://github.com/zatosource/zato/blob/main/code/zato-server/src/zato/server/service/store.py#L72

The point of this piece of code is that it adds connection or resource-related attribute directly to class_ which in this case is a class representing a service.

Now, you may wonder why this step of doing something like class_._out_plain_http = service_store.server.worker_store._worker_config.out_plain_http, why cannot we look up things directly in self._worker_config as you did it in PR #774.

The answer is that assigning attributes directly to class_ shaves off a significant part of the attribute lookup overhead.

Anything that happens in each service's __init__ method is part of the hottest path executed each time a service is invoked so even such tiny micro-optimizations ultimately amount to tangible outcome with the end result being that Zato 3.0 is several times faster than 2.0.

If you are interested, would you please be able to close PR #774 and instead add this attribute to WorkerStore as described above?

Many thanks!

**Comment by shizacat** [comment]
Thank you for your detailed explanation!

**Comment by shizacat** [comment]
@raybog Please close this issue, fixed in #776

**Comment by dsuch** [comment]
Done in 3d745c171855c938ce29e2358123a2e6a3aa3263, closing the ticket.
